### PR TITLE
Migrate .dockercfg to .docker/config.json and support for HTTP Headers

### PR DESCRIPTION
**Pull request by duglin**
This PR does the following:
- migrated ~/.dockerfg to ~/.docker/config.json. The data is migrated
  but the old file remains in case its needed
- moves the auth json in that fie into an "auth" property so we can add new
  top-level properties w/o messing with the auth stuff
- adds support for an HttpHeaders property in ~/.docker/config.json
  which adds these http headers to all msgs from the cli

In a follow-on PR I'll move the config file process out from under
"registry" since it not specific to that any more. I didn't do it here
because I wanted the diff to be smaller so people can make sure I didn't
break/miss any auth code during my edits.

Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

**Comment by duglin** [comment]
ping @jfrazelle @tiborvass @crosbymichael

**Comment by tiborvass** [comment]
@duglin it's interesting but I'm not sure I understand the usecase. What's the problem this is solving?

**Comment by duglin** [comment]
@tiborvass there are two things at play here:
1 - the more generic need for a config file that 3rd party tools can modify to change the behavior of the CLI. I think a 'DockerHost' property is the best example of this.  Some tooling (perhaps even Machine) needs to point the user's CLI to a particular docker host.  Modifying their env vars for them or asking them to do it, or use -H, isn't ideal.  So, a config file can be modified instead.  I would like to add "DockerHost" support but based on previous chats, for some reason, that might be controversial so I left if out for now.

2 - in our env we have a (sort of) proxy sitting in-front of a set of docker hosts. This proxy will do certain things including auth processing. We want our users to continue to use the default docker CLI to interact with our backends but in order for the auth stuff to work we need some additional auth http headers in all of the CLI messages.  So, we'll have some a tool (like a "login" exe) that the user to run to generate the proper auth token and then that tool will modify the config file to specify what http headers the docker cli should then include.  So, once the user is logged-in they can then use the Docker cli as they normally would.

**Comment by tianon** [comment]
Sorry for the premature code review move -- this still needs design review.

I'm +1 on the idea of `~/.docker/config.json`, but I personally think that we ought to do cascading config from `/etc/docker/config.json` to `~/.docker/config.json` to some command-line flags (like SSH does, for example), so I only see this as part of a bigger whole.  Do you see those kinds of things as simple additions later?

How do we prevent the community at large from building tools to parse and modify this file while we're at it, since there won't be any standard for ensuring that two tools don't try to write it at the same time before we get around to doing so?

**Comment by duglin** [comment]
I'm ok with both concepts (adding /etc/docker/config.json into the cascade && some kind of API/mechanism/something to help edit these files).  But I think both can be added later.  
Right now the CLI doesn't look at /etc/docker/... at all so that would a new concept and easy to add to the mix.  We don't even have any overlapping config data between ~/.docker/config.json and anything else yet :-)
New tooling/something to help edit the files can be added in a subsequent PR too.

**Comment by duglin** [comment]
@crosbymichael since this is a UX change, can you chime in here?

**Comment by tiborvass** [comment]
I'm +1 for this PR, as well as +1 to what tianon said.

@dmp42 @dmcgowan does this make sense to you too?

Ping @jfrazelle @icecrime

**Comment by duglin** [comment]
I'm ok with adding an `--httpHeader` cli flag if others are ok with it.  Thoughts?

**Comment by tiborvass** [comment]
The rationale for adding a flag is to make sure that all configs have an equivalent flags. Now @duglin told me that initially this was part of `.dockercfg`, which doesn't really have a flag and is really just auth-focused. So I'm not exactly sure what to think anymore about flag or not flag. Suggestions are welcome!

**Comment by dmcgowan** [comment]
Flags can always be added later when there is a real use for it. I am +1 to design, I think flags and cascading can be deferred until the use case is larger.  I am still uncomfortable with mixing in our existing auth configuration with general configuration though. The only reason I don't think that is a blocker is that we NEED to be moving away from the current way of storing auth and move to only storing credential identifiers (username only, or key fingerprints). In that case, the use case @duglin described around setting token headers and storing these identifiers are a good mix.

**Comment by duglin** [comment]
@dmcgowan I'm sorry if there's some confusion - or maybe I'm confused  :-)
This PR is different from #11759.  This one leaves .dockercfg alone but adds a new config file (.docker/config.json) to hold CLI options - like this new httpHeader info.
My plan is to split #11759 into two pieces of work.  Use this PR to create .docker/config.json and then morph #11759 into one that moves .dockercfg to be .docker/auth.json (or something like that).
My sense was that people wanted smaller/incremental changes.  
See the first 2 bullets in:  https://github.com/docker/docker/pull/11759#issuecomment-88482515   
I was going to leave the 3rd bullet for later since that's a much bigger piece of work and I'm not sure everyone is bought into that idea yet.
Are you ok with this or would you prefer a different approach?

**Comment by dmcgowan** [comment]
Sorry, my confusion, my initial skim failed to see one was using `registry.LoadConfig` and the other `LoadConfigFile`.  My +1 remains

**Comment by duglin** [comment]
are we ok with moving this to code review?

**Comment by tiborvass** [comment]
@jfrazelle what do you think about this?

**Comment by jfrazelle** [comment]
Sorry I read this a few times but I'm still confused about the flag, do we need it now or is it being added later if we need it

**Comment by duglin** [comment]
@jfrazelle I think that's up to @tiborvass and how much he wants it now.  @dmcgowan would prefer to wait. I'm ok either way. Do you have a preference ?

**Comment by jfrazelle** [comment]
ok wait sorry I read this all wrong... whoa this is just a config for flags right, so then i think we need the flag for consistency right

**Comment by duglin** [comment]
perhaps, it depends on whether we think people will want to add new http headers via the CLI command line or whether we view it more like a 3rd party tooling thing and they'll just use the config file.

**Comment by jfrazelle** [comment]
what is the usecase for the headers

**Comment by duglin** [comment]
@jfrazelle see https://github.com/docker/docker/pull/12009#issuecomment-88964710

**Comment by tiborvass** [comment]
Blame me for any confusion around flags :P

If you guys think we can add the cli flag later, then so be it :)

So we're talking about a config.json with only one key supported: `HttpHeaders`. It would have no flag equivalent in this PR.

**Comment by jfrazelle** [comment]
I think maybe lets add the flag later, I get nervous about flags, is this like an IBM specific proxy thing, I'm super confused still

**Comment by tiborvass** [comment]
@jfrazelle usecase is mostly for enterprises I guess. The PR itself is not IBM specific, it's generic enough in the sense that it provides a way for anyone to add custom HTTP Headers to all requests made by the client. And it's also generic in the sense that, that same config file can be reused for anything we want, even for auth once we move off of dockercfg.

@dmcgowan but you guys wouldn't use the HttpHeaders part of the config right? You would have maybe an Auth key in the config, or am I missing something?

**Comment by jfrazelle** [comment]
ok cool sounds fine to me, i feel like needs docs around a sample use case
for the headers but thats later

**Comment by duglin** [comment]
@jfrazelle its not IBM specific, in fact if you look at the openstack magnum project I believe they're looking at adding additional http headers to the message flow too ( https://github.com/openstack/magnum/blob/b1e166e22df7bafea1861b48a0fa549013e70c00/specs/containers-service.rst ).  I don't know if they will want to leverage this config file but it will be an option for them.

**Comment by jfrazelle** [comment]
ah ok cool thanks

**Comment by duglin** [comment]
Ok to move to code review?

**Comment by jfrazelle** [comment]
ping @crosbymichael

**Comment by duglin** [comment]
ping?

**Comment by duglin** [comment]
@crosbymichael you ok moving this to code review?

**Comment by crosbymichael** [comment]
After long discussions and intense battles lets do this.

On upgrade, migrate the old `.dockercfg` file to a subsection in `.docker/config.json` for registries and then ensure the the format is able to expand as needed with the additional headers as a first step.  Everything else in this PR looks good. 

Basically, the only change left to make in this PR is that we migrate the old config over if `config.json` does not exist and that's it.

**Comment by dmp42** [comment]
cc @docker/distribution-maintainers for review

**Comment by duglin** [comment]
I've updated the PR based on @crosbymichael's comment. Please give it another look.
/cc @tiborvass @tianon @dmcgowan

File: docs/sources/reference/commandline/cli.md [hunk 1]
```diff
@@ -48,6 +48,33 @@ These Go environment variables are case-insensitive. See the
 [Go specification](http://golang.org/pkg/net/http/) for details on these
 variables.
 
+## Configuration Files
+The Docker command line stores its configuration files in a directory
+called `.docker` within your home directory. Some of the files in there
+are managed by `docker` and should not be modified. However, there is
+a file called called `config.json` that can be modified to control certain
```
Comment (tiborvass) [inline docs/sources/reference/commandline/cli.md]: s/called called/called/

File: docs/sources/reference/commandline/cli.md [hunk 2]
```diff
@@ -48,6 +48,33 @@ These Go environment variables are case-insensitive. See the
...
+If there are options within `config.json` that modify the same behavior as
+an environment variable, or a command line option, then the order of
+precedence is: command line options override environment variables,
+which override `config.json` file properties.
+
+The `config.json` file is stored as a JSON encoding of the following set of
+properties:
+
+* HttpHeaders - a set of HTTP Headers that will be included in all
+  messages sent from the client to the deamon. Docker will not try
```
Comment (tiborvass) [inline docs/sources/reference/commandline/cli.md]: daemon

File: registry/auth.go [hunk 3]
```diff
@@ -147,9 +152,42 @@ func decodeAuth(authStr string) (string, string, error) {
...
+func LoadConfig(configDir string) (*ConfigFile, error) {
+	origConfigDir := configDir
+
+	if configDir == "" {
+		configDir = filepath.Join(homedir.Get(), ".docker")
+	}
+
+	configFile := ConfigFile{
+		AuthConfigs: make(map[string]AuthConfig),
+		filename:    filepath.Join(configDir, "config.json"),
```
Comment (dmcgowan) [inline registry/auth.go]: Is not using defined constant

File: integration-cli/docker_cli_config_test.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	"github.com/docker/docker/pkg/homedir"
+)
+
+func TestConfigHttpHeader(t *testing.T) {
+	testRequires(t, UnixCli) // Can' set/unset HOME on windows right now
```
Comment (dmcgowan) [inline integration-cli/docker_cli_config_test.go]: Small typo: missing the 't'

File: docs/sources/reference/commandline/cli.md (hunk 1 above)
Comment (duglin) [inline docs/sources/reference/commandline/cli.md]: Fixed

File: docs/sources/reference/commandline/cli.md (hunk 2 above)
Comment (duglin) [inline docs/sources/reference/commandline/cli.md]: Fixed

File: registry/auth.go (hunk 3 above)
Comment (duglin) [inline registry/auth.go]: Fixed

File: integration-cli/docker_cli_config_test.go (hunk 4 above)
Comment (duglin) [inline integration-cli/docker_cli_config_test.go]: Fixed

**Comment by dmcgowan** [comment]
In my testing existing keys from ~/.dockercfg were not getting migrated to ~/.docker/config.json after logging in to a new registry.  Then when attempting to push to a previously logged in registry I had to re-auth.

**Comment by duglin** [comment]
@dmcgowan hmmm ok - can you send me your config file so I can test with it?

**Comment by dmcgowan** [comment]
Not with my passwords :smile:
Also seems to be a bug when writing out a new entry to the "auths" key, the previously saved entries get their "auth" value reset to "Og==", which the base64 encoding of `:`.

To reproduce, have existing entries in your `.dockercfg`, login to a different registry (will be stored in `.docker/config.json`, then attempt to do an action which requires auth on a registry previously configured in `.dockercfg`.

For the second case, just login in to two different registries and you will see the previously saved gets overwritten.

**Comment by duglin** [comment]
@dmcgowan I found at least part of it - please try it again and let me know which parts (if any) still fail for you.  I'm pretty sure I at least have the "migration" part fixed, although I wasn't able to reproduce the "Og==" part.

**Comment by dmcgowan** [comment]
The migration works now, the "Og==" part is still happening. I will help you look into it more since I have the reproducible case.  Sorry for just dropping the problems and leaving yesterday, I'll see if I can track it down.

File: registry/auth.go [hunk 5]
```diff
@@ -147,18 +152,44 @@ func decodeAuth(authStr string) (string, string, error) {
...
+		filename:    filepath.Join(configDir, CONFIGFILE),
+	}
+
+	// Try happy path first - latest config file
+	if _, err := os.Stat(configFile.filename); err == nil {
+		data, err := ioutil.ReadFile(configFile.filename)
+		if err != nil {
+			return &configFile, err
+		}
+		err = json.Unmarshal(data, &configFile)
```
Comment (dmcgowan) [inline registry/auth.go]: The usernames/passwords are getting cleared out later because the config value is simply returned here without breaking apart the encoded values back into username and password. Later when the config is saved again, the Auth value is set, but Auth gets overridden using `encodeAuth(&authCopy)`, which encodes using the now empty username and password.

This probably needs to be done here for each value
```
authConfig.Username, authConfig.Password, err = decodeAuth(authConfig.Auth)
if err != nil {
	return &configFile, err
}
authConfig.Auth = ""
```

Comment (duglin) [inline registry/auth.go]: @dmcgowan thanks a ton!  I totally forgot that the data was encoded.  Its fixed now and made sure the testcases covered this case.

**Comment by duglin** [comment]
@dmcgowan let me know if you still see weird behavior.

**Comment by dmcgowan** [comment]
Fixed, LGTM

**Comment by duglin** [comment]
any other comments?
@icecrime @tianon

**Comment by thaJeztah** [comment]
Hm  design review has passed, but should the config-file format be versioned (some meta-data)? This could make it easier to take future migrations into account.

**Comment by duglin** [comment]
not sure that's really necessary given json is extensible and we can just drop/add/migrate fields as needed.

**Comment by thaJeztah** [comment]
It may be overkill (at least, now), just that I've seen many occasions in this project where "guessing" is involved. Just a suggestion though, no hard feelings if not :)

**Comment by endophage** [comment]
Without versioning you can only add fields. Migrating fields relies on knowing what you are migrating from and to. Dropping fields assumes an old version will never require that field to be present.

I'd always vote to have a version number. Adding it later only solves going forward and once you decide it's necessary, it's too late to retroactively fix.

**Comment by duglin** [comment]
Remember, we’re not talking about a protocol thing, we’re talking about a config file which is more like an internal data structure for use by Docker. I don’t think there’s a requirement for a migrated config file to be used by an old version of Docker.

**Comment by endophage** [comment]
I guess it depends on whether there is any expectation for a config file to be distributed separately from the application consuming it. Might a company want to preload new employee computers with a config file they keep separately to the version of Docker they are installing? A version number might at least provide them a clear notification should the config and application not be compatible.

I'm generally in favour of versioning because it takes just a dozen or so bytes of data and can enable more flexibility in the future should there ever be a reason to create breaking changes. It's a dozen bytes of insurance, we all hope it's never needed, and we're all glad we have it when it is.

File: registry/auth.go [hunk 6]
```diff
@@ -179,48 +224,48 @@ func LoadConfig(rootPath string) (*ConfigFile, error) {
...
-		return nil
-	}
-
-	configs := make(map[string]AuthConfig, len(configFile.Configs))
-	for k, authConfig := range configFile.Configs {
+func (configFile *ConfigFile) Save() error {
+	// Encode sensitive data into a new/temp stuct
+	tmpAuthConfigs := make(map[string]AuthConfig, len(configFile.AuthConfigs))
+	for k, authConfig := range configFile.AuthConfigs {
 		authCopy := authConfig
```
Comment (tiborvass) [inline registry/auth.go]: are you sure this will do a copy?

File: registry/auth.go [hunk 7]
```diff
@@ -179,48 +224,48 @@ func LoadConfig(rootPath string) (*ConfigFile, error) {
...
+			configFile.AuthConfigs[k] = authConfig
 		}
 	}
 	return &configFile, nil
 }
 
-// save the auth config
-func SaveConfig(configFile *ConfigFile) error {
-	confFile := path.Join(configFile.rootPath, CONFIGFILE)
-	if len(configFile.Configs) == 0 {
```
Comment (tiborvass) [inline registry/auth.go]: @duglin why remove this part?

Comment (duglin) [inline registry/auth.go]: That code would remove the file if the auth stuff was empty. That really only made sense when the only thing in there was the auth stuff. Now that we can have other things this check doesn't make sense.

File: registry/auth.go [hunk 8]
```diff
@@ -179,48 +224,48 @@ func LoadConfig(rootPath string) (*ConfigFile, error) {
...
-		os.Remove(confFile)
-		return nil
-	}
-
-	configs := make(map[string]AuthConfig, len(configFile.Configs))
-	for k, authConfig := range configFile.Configs {
+func (configFile *ConfigFile) Save() error {
+	// Encode sensitive data into a new/temp stuct
```
Comment (tiborvass) [inline registry/auth.go]: `struct`

File: api/client/cli.go [hunk 9]
```diff
@@ -167,9 +160,15 @@ func NewDockerCli(in io.ReadCloser, out, err io.Writer, keyFile string, proto, a
 		tr.Dial = (&net.Dialer{Timeout: timeout}).Dial
 	}
 
+	configFile, e := registry.LoadConfig(filepath.Join(homedir.Get(), ".docker"))
+	if e != nil {
+		fmt.Fprintf(err, "WARNING: Error loading config file:%s\n", e)
```
Comment (tiborvass) [inline api/client/cli.go]: %v

File: api/client/utils.go [hunk 10]
```diff
@@ -299,7 +306,7 @@ func (cli *DockerCli) monitorTtySize(id string, isExec bool) error {
 		sigchan := make(chan os.Signal, 1)
 		gosignal.Notify(sigchan, signal.SIGWINCH)
 		go func() {
-			for _ = range sigchan {
+			for range sigchan {
```
Comment (tiborvass) [inline api/client/utils.go]: go 1.4ism, I would revert this line just in case

File: api/client/cli.go (hunk 9 above)
Comment (duglin) [inline api/client/cli.go]: done

File: registry/auth.go (hunk 8 above)
Comment (duglin) [inline registry/auth.go]: done

File: api/client/utils.go (hunk 10 above)
Comment (duglin) [inline api/client/utils.go]: done - always hits me on this file.

File: integration-cli/docker_cli_config_test.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+	defer func() { os.Setenv(homeKey, homeVal) }()
+	os.Setenv(homeKey, tmpDir)
+
+	data := `{
+		"HttpHeaders": { "MyHeader": "MyValue" }
+	}`
+
+	err := ioutil.WriteFile(tmpCfg, []byte(data), 0600)
+	if err != nil {
+		t.Fatalf("Err creating file(%s): %s", tmpCfg, err)
```
Comment (tiborvass) [inline integration-cli/docker_cli_config_test.go]: %v for err

File: registry/auth.go [hunk 12]
```diff
@@ -147,18 +152,58 @@ func decodeAuth(authStr string) (string, string, error) {
...
+	}
+
+	// Try happy path first - latest config file
+	if _, err := os.Stat(configFile.filename); err == nil {
+		data, err := ioutil.ReadFile(configFile.filename)
+		if err != nil {
+			return &configFile, err
+		}
+
+		err = json.Unmarshal(data, &configFile)
```
Comment (tiborvass) [inline registry/auth.go]: use os.Open and json.NewDecoder

File: api/client/utils.go (hunk 10 above)
Comment (tiborvass) [inline api/client/utils.go]: @LK4D4 confirmed we're okay for dropping support for go1.3, so let's leave it.

File: registry/auth.go [hunk 13]
```diff
@@ -147,18 +152,58 @@ func decodeAuth(authStr string) (string, string, error) {
...
+		if err != nil {
+			return &configFile, err
+		}
+
+		for addr, ac := range configFile.AuthConfigs {
+			ac.Username, ac.Password, err = decodeAuth(ac.Auth)
+			if err != nil {
+				return &configFile, err
```
Comment (tiborvass) [inline registry/auth.go]: what's the usecase for returning configFile upon error? I'd prefer returning `nil, err`

File: registry/auth.go [hunk 14]
```diff
@@ -147,18 +152,58 @@ func decodeAuth(authStr string) (string, string, error) {
...
 		return &configFile, nil //missing file is not an error
 	}
+
 	b, err := ioutil.ReadFile(confFile)
 	if err != nil {
 		return &configFile, err
 	}
 
-	if err := json.Unmarshal(b, &configFile.Configs); err != nil {
+	if err := json.Unmarshal(b, &configFile.AuthConfigs); err != nil {
```
Comment (tiborvass) [inline registry/auth.go]: os.Open + json.NewDecoder

File: integration-cli/docker_cli_config_test.go (hunk 11 above)
Comment (duglin) [inline integration-cli/docker_cli_config_test.go]: done

File: registry/config_file_test.go [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/docker/docker/pkg/homedir"
+)
+
+func pass() {
+	pc := make([]uintptr, 10)
+	runtime.Callers(0, pc)
```
Comment (tiborvass) [inline registry/config_file_test.go]: wow so fancy, what are you trying to do ?

Comment (tiborvass) [inline registry/config_file_test.go]: refactor code and make sure the error line is from the caller?

File: registry/auth.go (hunk 12 above)
Comment (duglin) [inline registry/auth.go]: done

File: api/client/utils.go (hunk 10 above)
Comment (duglin) [inline api/client/utils.go]: reverted

File: registry/auth.go (hunk 13 above)
Comment (duglin) [inline registry/auth.go]: This was how the old code worked so I followed their lead.
If I had to guess I'd say it was because they didn't want to consider parsing the config file to be something worth stopping on.  We can change it but I'd prefer to do it in a separate PR because I have a sense it'll lead to a long bikeshedding discussion.  At least this PR doesn't change today's behavior.

File: registry/auth.go (hunk 14 above)
Comment (duglin) [inline registry/auth.go]: done

Comment (duglin) [inline registry/auth.go]: I take that back - I didn't make this change.
If you look below you'll see that if we can't parse the file then we assume we're dealing with a really old file that (I think) isn't even json, notice the "Split" below.  In order to be able to do that we need to read in the entire file.
Given this is an old file I think we're safe to assume that its small enough to be read into memory safely and therefore we don't need to worry about the streaming json decoder stuff.  Of course, we may be able to remove this entire section too if we think we can drop support for the really old format.

File: registry/config_file_test.go (hunk 15 above)
Comment (duglin) [inline registry/config_file_test.go]: :-)    Sort of.  Was tied of trying to come up with descriptive text for testcases when I figured the names of the test should do a good job of it.  So, I decided to see what it would look like if we used reflection to print the "pass" message.  If people like this I can dup it for the integration-cli tests.

File: registry/auth.go (hunk 14 above)
Comment (tiborvass) [inline registry/auth.go]: fair enough

**Comment by tiborvass** [comment]
It's super weird, but i did a `docker login` and logged in to the hub, and I have this:

```
root@68d34a6eb689:/go/src/github.com/docker/docker# docker login
Username: tiborvass
Password:
Email: e@e.com
WARNING: login credentials saved in /root/.docker/config.json.
Login Succeeded
root@68d34a6eb689:/go/src/github.com/docker/docker# cat /root/.docker/config.json
cat: /root/.docker/config.json: No such file or directory
root@68d34a6eb689:/go/src/github.com/docker/docker# find / -name config.json
root@68d34a6eb689:/go/src/github.com/docker/docker#
```

**Comment by duglin** [comment]
hmmm, I just rebuilt and I don't see the issue. I get the file in the right spot.

**Comment by duglin** [comment]
Added checks to return an error from Save() maybe that'll show what's going on for @tiborvass

**Comment by tiborvass** [comment]
Was missing an mkdir for .docker

**Comment by duglin** [comment]
fixed

**Comment by tiborvass** [comment]
@duglin I copied my .dockercfg and did a docker login, and it didn't pick up the old config (otherwise it would show my username in parentheses).

**Comment by tiborvass** [comment]
LGTM

**Comment by tiborvass** [comment]
@duglin actually, i suggest we remove .dockercfg if config.json has been successfully written

**Comment by dmcgowan** [comment]
@tibor can we phase a change like that in?  If we removed it then its going to be pain switching between versions for awhile

**Comment by duglin** [comment]
@tiborvass you ok with @dmcgowan's suggestion?

**Comment by tiborvass** [comment]
@dmcgowan then we have to make sure we write to both files.

**Comment by tiborvass** [comment]
@duglin @dmcgowan I also suggest that in debug mode, we output when we write to the old config file in addition to when we write to the new config file.

**Comment by duglin** [comment]
I don't think we write to both files. I don't mind keeping the old file around either for old clients or in case there's a bug in the new code (although its just userid/passwd's - not hard to recreate).  But I do agree with @tiborvass that it should be short lived - perhaps before 1.7 ships?  @dmcgowan thoughts?

**Comment by duglin** [comment]
rebased
ping @icecrime @LK4D4

**Comment by estesp** [comment]
reviewed code and tested locally with prior login details stored; all worked properly

LGTM - assume we need docs review, and a decision on @dmcgowan @duglin discussion above regarding dual-support of old and new config files for a time.

File: registry/config_file_test.go (hunk 15 above)
Comment (LK4D4) [inline registry/config_file_test.go]: Hmm, actually `go test -v` shows `PASS` messages with Test names. I'm not sure for what we need this function.

Comment (duglin) [inline registry/config_file_test.go]: does `make test` run it with `-v` ?  make I do `make test` I only see the print stuff from this func, nothing else.

Comment (LK4D4) [inline registry/config_file_test.go]: Yeah, but it looks inconsistent. If we want this for all tests, we can use `-v`. But why we want it only for one package?

Comment (duglin) [inline registry/config_file_test.go]: see above - I was just playing around and didn't want to add a printf for each test. But if you feel strongly about it I can remove it.   I also wouldn't mind doing it for all tests but that's a different topic.

Comment (LK4D4) [inline registry/config_file_test.go]: But doing this for all tests is simply add `-v` for `go test`. Yeah, I'm not in favor of some output in middle of test-suite for now. And personally I'm not sure why I want to see passed messages at all, but this is discussion for other PR.

**Comment by LK4D4** [comment]
LGTM
apart from `pass()` function

**Comment by duglin** [comment]
Despite it being cool :-)  I remove pass() just for @LK4D4

File: registry/config_file_test.go [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+	js := ` { "auths": { "https://index.docker.io/v1/": { "auth": "am9lam9lOmhlbGxv", "email": "joe@gmail.com" } } }`
+	ioutil.WriteFile(fn, []byte(js), 0600)
+
+	config, err := LoadConfig(tmpHome)
+	if err != nil {
+		t.Fatalf("Failed loading on empty json file: %q", err)
+	}
+
+	ac := config.AuthConfigs["https://index.docker.io/v1/"]
+	if ac.Email != "joe@gmail.com" || ac.Username != "joejoe" || ac.Password != "hello" {
```
Comment (tiborvass) [inline registry/config_file_test.go]: @duglin huh just saw this, let's use `user@example.org` instead.

Comment (duglin) [inline registry/config_file_test.go]: updated

File: docs/sources/reference/commandline/cli.md [hunk 17]
```diff
@@ -48,6 +48,34 @@ These Go environment variables are case-insensitive. See the
 [Go specification](http://golang.org/pkg/net/http/) for details on these
 variables.
 
+## Configuration Files
+The Docker command line stores its configuration files in a directory
```
Comment (moxiegirl) [inline docs/sources/reference/commandline/cli.md]: ## Configuration Files

The Docker command line stores its configuration files in a directory called
`.docker` within your `HOME` directory. Docker manages most of the files in
`.docker` and you should not modify them. However, you *can modify* the
`.docker/config.json` file to control certain aspects of how the `docker`
command behaves.

Currently, you can modify the `docker` command behavior using environment variables or command-line options. You can also use options within `config.json` to modify that same behavior.  When using these mechanisms, you must keep in mind the order of precedence among them. Command line options override environment variables and environment variables override properties you specify in a `config.json` file.

The `config.json` file stores a JSON encoding of a single `HttpHeaders`
property. The property specifies a set of headers to include in all
messages sent from the Docker client to the daemon. Docker does not try to
interpret or understand these header; it simply puts them into the messages.
Docker does not allow these headers to change any headings it sets for itself.

Following is a sample `config.json` file:

    {
      "HttpHeaders: {
        "MyHeader": "MyValue"
      }
    }

**Comment by moxiegirl** [comment]
LGTM

ping @docker/docs-owners

**Comment by fredlf** [comment]
Docs LGTM with @moxiegirl 's revision.

File: docs/sources/reference/commandline/cli.md (hunk 17 above)
Comment (duglin) [inline docs/sources/reference/commandline/cli.md]: @moxiegirl ok - picked up your edits with just a minor tweak to the start of the 2nd paragraph. See if you're ok with that.

Comment (duglin) [inline docs/sources/reference/commandline/cli.md]: oh I also changed "headings" to "headers" in the last para.

File: registry/config_file_test.go [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+	defer os.RemoveAll(tmpHome)
+
+	homeKey := homedir.Key()
+	homeVal := homedir.Get()
+
+	defer func() { os.Setenv(homeKey, homeVal) }()
+	os.Setenv(homeKey, tmpHome)
+
+	fn := filepath.Join(tmpHome, OLD_CONFIGFILE)
+	js := `{"https://index.docker.io/v1/":{"auth":"am9lam9lOmhlbGxv","email":"@example.com"}}`
```
Comment (thaJeztah) [inline registry/config_file_test.go]: @duglin Looks like you created an incomplete email address after changing it from `@gmail.com`

Not sure it's a problem here, just noting it, just in case :)

Comment (duglin) [inline registry/config_file_test.go]: Not sure how I managed to do that but I fixed it. thanks!

**Comment by moxiegirl** [comment]
LGTM

File: registry/config_file_test.go (hunk 18 above)
Comment (thaJeztah) [inline registry/config_file_test.go]: Copy-pasta-slippo :smile: :+1:

File: registry/auth.go [hunk 19]
```diff
@@ -179,48 +224,52 @@ func LoadConfig(rootPath string) (*ConfigFile, error) {
...
+	configFile.AuthConfigs = tmpAuthConfigs
+	defer func() { configFile.AuthConfigs = saveAuthConfigs }()
+
+	data, err := json.MarshalIndent(configFile, "", "\t")
 	if err != nil {
 		return err
 	}
-	err = ioutil.WriteFile(confFile, b, 0600)
+
+	if err := os.MkdirAll(filepath.Dir(configFile.filename), 0600); err != nil {
```
Comment (dmcgowan) [inline registry/auth.go]: Argh was still testing this fix when the merge came in.  This permission causing problems and needs to be 0700.

Comment (duglin) [inline registry/auth.go]: ok - I can easily change that.  But I'm curious, why do you need +x ?

Comment (dmcgowan) [inline registry/auth.go]: On the directory

Comment (dmcgowan) [inline registry/auth.go]: Without the x bit traversing will still throw a permission, writing to "~/.docker/config.json" failed and even attempting to touch is not allowed.

~~~
$ ls -ld ~/.docker
drw-------. 2 derek derek 4096 Apr 20 12:18 /home/derek/.docker
$ touch ~/.docker/config.json
touch: cannot touch ‘/home/derek/.docker/config.json’: Permission denied

~~~

Comment (duglin) [inline registry/auth.go]: That's what I would have expected. However, if you look at this branch: https://github.com/duglin/docker/tree/TweakConfigPerms
I added a testcase to have Save() create the dir and when I left the code at 0600 it still worked - even was able to read the new config file.  Is the next test wrong?

Comment (duglin) [inline registry/auth.go]: could it be because the test is running as root? dunno if it is or not, just speculating.
